### DISCUSSION TITLE
Open/Close vs Start/Shutdown

### DISCUSSION THREAD

**Discussion by okbaswc** [microsoft/msquic#1180] [Q&A] [answered]
I'm a little confused how best to cleanup on application exit.

1.) If my server wants to start a unidirectional stream, do I call "StreamOpen" followed immediately by "StreamStart"?
2.) If my server started a unidirectional stream with client and I want to destroy that stream, do I need to call "StreamShutdown" followed immediately by "StreamClose"?
3.) If the peer started a stream and I want to exit the server application, do I need to call "StreamClose" or "StreamShutdown" at all?
4.) Likewise, on server application exit, should I call "ConnectionShutdown" or "ConnectionClose"?

**Comment by nibanks** [comment, marked as answer]
Have you seen our API docs? https://github.com/microsoft/msquic/blob/main/docs/API.md

Generally, you can think of Close functions as `free` calls on those objects, and therefore must be the last thing you call. Shutdown starts the clean up process, informing the peer of what's going on. As for Stream start, yes that must be called on every stream object after you open it.
